File: net/sync/service/src/main/java/de/danoeh/antennapod/net/sync/service/podhopper/SupabaseClient.java
package de.danoeh.antennapod.net.sync.service.podhopper;

import androidx.annotation.NonNull;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import de.danoeh.antennapod.net.sync.serviceinterface.SyncServiceException;
import de.danoeh.antennapod.storage.preferences.SynchronizationCredentials;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class SupabaseClient {
    private static final MediaType JSON_TYPE = MediaType.parse("application/json; charset=utf-8");
    private static final long TOKEN_SAFETY_MARGIN_MS = 60 * 1000;

    private static String cachedAccessToken = null;
    private static String cachedUserId = null;
    private static long tokenExpiresAtMs = 0;

    private final OkHttpClient httpClient;
    private final String email;
    private final String password;

    public SupabaseClient(@NonNull OkHttpClient httpClient, String email, String password) {
        this.httpClient = httpClient;
        this.email = email;
        this.password = password;
    }

    public synchronized String ensureSession() throws SyncServiceException {
        if (cachedAccessToken != null && System.currentTimeMillis() < tokenExpiresAtMs - TOKEN_SAFETY_MARGIN_MS) {
            return cachedAccessToken;
        }
        String refreshToken = SynchronizationCredentials.getRefreshToken();
        if (password == null && refreshToken != null) {
            return refreshSession(refreshToken);
        }
        if (email == null || password == null) {
            throw new SyncServiceException("PodHopper sync: not logged in");
        }
        try {
            JSONObject body = new JSONObject();
            body.put("email", email);
            body.put("password", password);
            JSONObject json = authCall("/auth/v1/token?grant_type=password", body);
            applySession(json);
            return cachedAccessToken;
        } catch (JSONException e) {
            throw new SyncServiceException(e);
        }
    }

    private synchronized String refreshSession(String refreshToken) throws SyncServiceException {
        try {
            JSONObject body = new JSONObject();
            body.put("refresh_token", refreshToken);
            JSONObject json;
            try {
                json = authCall("/auth/v1/token?grant_type=refresh_token", body);
            } catch (AuthRejectedException e) {
                SynchronizationCredentials.setRefreshToken(null);
                throw new SyncServiceException("PodHopper pairing expired. Re-pair this device with your phone.");
            }
            applySession(json);
            String rotated = json.optString("refresh_token", null);
            if (rotated != null) {
                SynchronizationCredentials.setRefreshToken(rotated);
            }
            return cachedAccessToken;
        } catch (JSONException e) {
            throw new SyncServiceException(e);
        }
    }

    public boolean signUp(String signupEmail, String signupPassword) throws SyncServiceException {
        try {
            JSONObject body = new JSONObject();
            body.put("email", signupEmail);
            body.put("password", signupPassword);
            JSONObject json = authCall("/auth/v1/signup", body);
            if (json.has("access_token")) {
                applySession(json);
                return true;
            }
            return false;
        } catch (JSONException e) {
            throw new SyncServiceException(e);
        }
    }

    public void recoverPassword(String recoveryEmail) throws SyncServiceException {
        try {
            JSONObject body = new JSONObject();
            body.put("email", recoveryEmail);
            authCall("/auth/v1/recover", body);
        } catch (JSONException e) {
            throw new SyncServiceException(e);
        }
    }

    public static synchronized String claimPairingSession(OkHttpClient client, String tokenHash)
            throws SyncServiceException {
        try {
            JSONObject body = new JSONObject();
            body.put("type", "magiclink");
            body.put("token_hash", tokenHash);
            JSONObject json = staticAuthCall(client, "/auth/v1/verify", body);
            cachedAccessToken = json.getString("access_token");
            long expiresInSec = json.optLong("expires_in", 3600);
            tokenExpiresAtMs = System.currentTimeMillis() + expiresInSec * 1000;
            JSONObject user = json.getJSONObject("user");
            cachedUserId = user.getString("id");
            String rotated = json.optString("refresh_token", null);
            if (rotated != null) {
                SynchronizationCredentials.setRefreshToken(rotated);
            }
            return user.optString("email", "");
        } catch (JSONException e) {
            throw new SyncServiceException(e);
        }
    }

    private synchronized void applySession(JSONObject json) throws JSONException {
        cachedAccessToken = json.getString("access_token");
        long expiresInSec = json.optLong("expires_in", 3600);
        tokenExpiresAtMs = System.currentTimeMillis() + expiresInSec * 1000;
        cachedUserId = json.getJSONObject("user").getString("id");
    }

    private JSONObject authCall(String path, JSONObject body) throws SyncServiceException {
        return staticAuthCall(httpClient, path, body);
    }

    private static JSONObject staticAuthCall(OkHttpClient client, String path, JSONObject body)
            throws SyncServiceException {
        Request request = new Request.Builder()
                .url(PodHopperConfig.SUPABASE_URL + path)
                .addHeader("apikey", PodHopperConfig.SUPABASE_ANON_KEY)
                .post(RequestBody.create(body.toString(), JSON_TYPE))
                .build();
        try (Response response = client.newCall(request).execute()) {
            String responseBody = response.body() != null ? response.body().string() : "";
            if (response.code() == 400 || response.code() == 401 || response.code() == 403) {
                throw new AuthRejectedException("PodHopper auth failed: HTTP "
                        + response.code() + " " + responseBody);
            }
            if (!response.isSuccessful()) {
                throw new SyncServiceException("PodHopper auth failed: HTTP "
                        + response.code() + " " + responseBody);
            }
            if (responseBody.isEmpty()) {
                return new JSONObject();
            }
            return new JSONObject(responseBody);
        } catch (IOException | JSONException e) {
            throw new SyncServiceException(e);
        }
    }

    public static class AuthRejectedException extends SyncServiceException {
        private static final long serialVersionUID = 1L;

        public AuthRejectedException(String message) {
            super(message);
        }
    }

    public synchronized String getUserId() throws SyncServiceException {
        ensureSession();
        return cachedUserId;
    }

    public static synchronized void clearSession() {
        cachedAccessToken = null;
        cachedUserId = null;
        tokenExpiresAtMs = 0;
    }

    public void upsert(String table, String onConflictColumns, JSONArray rows) throws SyncServiceException {
        if (rows.length() == 0) {
            return;
        }
        String url = PodHopperConfig.SUPABASE_URL + "/rest/v1/" + table
                + "?on_conflict=" + onConflictColumns;
        Request request = restRequestBuilder(url)
                .addHeader("Prefer", "resolution=merge-duplicates,return=minimal")
                .post(RequestBody.create(rows.toString(), JSON_TYPE))
                .build();
        executeExpectingSuccess(request, true);
    }

    public JSONArray select(String table, String query) throws SyncServiceException {
        String url = PodHopperConfig.SUPABASE_URL + "/rest/v1/" + table + "?" + query;
        Request request = restRequestBuilder(url).get().build();
        String body = executeExpectingSuccess(request, true);
        try {
            return new JSONArray(body);
        } catch (JSONException e) {
            throw new SyncServiceException(e);
        }
    }

    private Request.Builder restRequestBuilder(String url) throws SyncServiceException {
        String token = ensureSession();
        return new Request.Builder()
                .url(url)
                .addHeader("apikey", PodHopperConfig.SUPABASE_ANON_KEY)
                .addHeader("Authorization", "Bearer " + token)
                .addHeader("Content-Type", "application/json");
    }

    private String executeExpectingSuccess(Request request, boolean retryOnAuthError) throws SyncServiceException {
        try (Response response = httpClient.newCall(request).execute()) {
            String responseBody = response.body() != null ? response.body().string() : "";
            if (response.code() == 401 && retryOnAuthError) {
                clearSession();
                ensureSession();
                Request retried = request.newBuilder()
                        .removeHeader("Authorization")
                        .addHeader("Authorization", "Bearer " + cachedAccessToken)
                        .build();
                return executeExpectingSuccess(retried, false);
            }
            if (!response.isSuccessful()) {
                throw new SyncServiceException("PodHopper sync request failed: HTTP "
                        + response.code() + " " + responseBody);
            }
            return responseBody;
        } catch (IOException e) {
            throw new SyncServiceException(e);
        }
    }
}

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/podhopper/SupabaseClient.kt
package au.com.shiftyjelly.pocketcasts.repositories.podhopper

import au.com.shiftyjelly.pocketcasts.preferences.Settings
import au.com.shiftyjelly.pocketcasts.servers.di.Raw
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

/**
 * Talks to PodHopper's Supabase backend: GoTrue auth plus PostgREST upsert and select.
 *
 * This is a faithful Kotlin port of the proven AntennaPod implementation. The session
 * (access token, user id, expiry) is cached in memory. The refresh token and signed-in email
 * are persisted in app settings, so the user stays signed in across restarts and the access
 * token is silently refreshed when it expires.
 *
 * The @Raw OkHttp client is used deliberately: it carries no Pocket Casts auth interceptor,
 * since Supabase supplies its own apikey and bearer token.
 */
@Singleton
class SupabaseClient @Inject constructor(
    @Raw private val httpClient: OkHttpClient,
    private val settings: Settings,
) {
    private val jsonType = "application/json; charset=utf-8".toMediaType()

    private var cachedAccessToken: String? = null
    private var cachedUserId: String? = null
    private var tokenExpiresAtMs: Long = 0

    fun isLoggedIn(): Boolean = settings.podhopperRefreshToken.value.isNotEmpty()

    /**
     * Emits the current sign-in state and every later change to it: true once a refresh token is
     * stored (sign-in or sign-up succeeded), false when it is cleared (sign-out or an expired
     * session). Lets sync components react the instant sign-in completes instead of polling
     * isLoggedIn() or racing an activity lifecycle callback.
     */
    val loginState: Flow<Boolean>
        get() = settings.podhopperRefreshToken.flow.map { it.isNotEmpty() }

    @Synchronized
    fun login(email: String, password: String) {
        val body = JSONObject()
        body.put("email", email)
        body.put("password", password)
        val json = authCall("/auth/v1/token?grant_type=password", body)
        applySession(json)
        settings.podhopperRefreshToken.set(json.optString("refresh_token", ""), updateModifiedAt = false)
        settings.podhopperEmail.set(email, updateModifiedAt = false)
    }

    /**
     * Creates a new account. When the project does not require email confirmation, Supabase returns
     * a session immediately: that session is applied and persisted exactly like [login], and this
     * returns true. When confirmation is required, no token comes back and this returns false, which
     * the caller surfaces as "check your email to confirm".
     */
    @Synchronized
    fun signUp(email: String, password: String): Boolean {
        val body = JSONObject()
        body.put("email", email)
        body.put("password", password)
        val json = authCall("/auth/v1/signup", body)
        if (json.has("access_token")) {
            applySession(json)
            settings.podhopperRefreshToken.set(json.optString("refresh_token", ""), updateModifiedAt = false)
            settings.podhopperEmail.set(email, updateModifiedAt = false)
            return true
        }
        return false
    }

    /**
     * Asks Supabase to send a password reset email. Succeeds quietly; failures throw.
     */
    fun recoverPassword(email: String) {
        val body = JSONObject()
        body.put("email", email)
        authCall("/auth/v1/recover", body)
    }

    @Synchronized
    fun ensureSession(): String {
        val cached = cachedAccessToken
        if (cached != null && System.currentTimeMillis() < tokenExpiresAtMs - TOKEN_SAFETY_MARGIN_MS) {
            return cached
        }
        val refreshToken = settings.podhopperRefreshToken.value
        if (refreshToken.isNotEmpty()) {
            return refreshSession(refreshToken)
        }
        throw SupabaseException("PodHopper sync: not signed in")
    }

    @Synchronized
    fun getUserId(): String? {
        ensureSession()
        return cachedUserId
    }

    @Synchronized
    fun logout() {
        clearSessionCache()
        settings.podhopperRefreshToken.set("", updateModifiedAt = false)
        settings.podhopperEmail.set("", updateModifiedAt = false)
    }

    fun upsert(table: String, onConflictColumns: String, rows: JSONArray) {
        if (rows.length() == 0) {
            return
        }
        val url = PodHopperConfig.SUPABASE_URL + "/rest/v1/" + table + "?on_conflict=" + onConflictColumns
        val request = restRequestBuilder(url)
            .addHeader("Prefer", "resolution=merge-duplicates,return=minimal")
            .post(rows.toString().toRequestBody(jsonType))
            .build()
        executeExpectingSuccess(request, retryOnAuthError = true)
    }

    fun select(table: String, query: String): JSONArray {
        val url = PodHopperConfig.SUPABASE_URL + "/rest/v1/" + table + "?" + query
        val request = restRequestBuilder(url).get().build()
        val body = executeExpectingSuccess(request, retryOnAuthError = true)
        return JSONArray(body)
    }

    /**
     * Approves a car or TV pairing code on behalf of the signed-in user. The other device (the car)
     * generates the code and polls for approval; this call, authenticated as the user, tells the
     * pairing edge function to grant that code access to the user's account. Blocking; call off the
     * main thread. Throws on failure (invalid or expired code, or network error).
     */
    fun approveCarPairing(code: String) {
        val url = PodHopperConfig.SUPABASE_URL + "/functions/v1/pairing"
        val body = JSONObject()
        body.put("action", "approve")
        body.put("code", code)
        val request = restRequestBuilder(url)
            .post(body.toString().toRequestBody(jsonType))
            .build()
        executeExpectingSuccess(request, retryOnAuthError = true)
    }

    /**
     * Car side of pairing, step 1. Asks the pairing edge function for a fresh code that the car
     * displays. Authenticated with the anon key only (the car is not signed in yet). Blocking; call
     * off the main thread. Returns the code string.
     */
    fun startPairing(deviceName: String): String {
        val body = JSONObject()
        body.put("action", "start")
        body.put("device_name", deviceName)
        val json = pairingCall(body)
        return json.getString("code")
    }

    /**
     * Car side of pairing, step 2. Polls the pairing edge function for the status of the code the
     * car is showing. Returns the token_hash once the phone has approved, null while still pending,
     * and throws when the code has expired. Authenticated with the anon key only. Blocking; call off
     * the main thread.
     */
    fun pollPairing(code: String): String? {
        val body = JSONObject()
        body.put("action", "poll")
        body.put("code", code)
        val json = pairingCall(body)
        val status = json.optString("status", "pending")
        if (status == "approved") {
            return json.getString("token_hash")
        }
        if (status == "expired") {
            throw SupabaseException("PodHopper pairing code expired")
        }
        return null
    }

    /**
     * Car side of pairing, step 3. Exchanges the approved token_hash for a real session through
     * Supabase Auth (the magiclink verify endpoint, anon key, no bearer), then applies and persists
     * that session exactly like a normal sign-in. Storing the refresh token is what flips
     * [isLoggedIn] to true and turns sync on. Blocking; call off the main thread.
     */
    @Synchronized
    fun claimPairingSession(tokenHash: String) {
        clearSessionCache()
        val body = JSONObject()
        body.put("type", "magiclink")
        body.put("token_hash", tokenHash)
        val json = authCall("/auth/v1/verify", body)
        applySession(json)
        val refreshToken = json.optString("refresh_token", "")
        if (refreshToken.isNotEmpty()) {
            settings.podhopperRefreshToken.set(refreshToken, updateModifiedAt = false)
        }
        val email = json.optJSONObject("user")?.optString("email").orEmpty()
        if (email.isNotEmpty()) {
            settings.podhopperEmail.set(email, updateModifiedAt = false)
        }
    }

    /**
     * Posts to the pairing edge function with the anon key as both apikey and bearer (the calls the
     * car makes before it has a user session). Throws on a non-success HTTP status.
     */
    private fun pairingCall(body: JSONObject): JSONObject {
        val request = Request.Builder()
            .url(PodHopperConfig.SUPABASE_URL + "/functions/v1/pairing")
            .addHeader("apikey", PodHopperConfig.SUPABASE_ANON_KEY)
            .addHeader("Authorization", "Bearer " + PodHopperConfig.SUPABASE_ANON_KEY)
            .post(body.toString().toRequestBody(jsonType))
            .build()
        httpClient.newCall(request).execute().use { response ->
            val responseBody = response.body.string()
            if (!response.isSuccessful) {
                throw SupabaseException("PodHopper pairing request failed: HTTP ${response.code} $responseBody")
            }
            return if (responseBody.isEmpty()) JSONObject() else JSONObject(responseBody)
        }
    }

    @Synchronized
    private fun refreshSession(refreshToken: String): String {
        val body = JSONObject()
        body.put("refresh_token", refreshToken)
        val json = try {
            authCall("/auth/v1/token?grant_type=refresh_token", body)
        } catch (e: AuthRejectedException) {
            settings.podhopperRefreshToken.set("", updateModifiedAt = false)
            throw SupabaseException("PodHopper sign-in expired. Sign in again.", e)
        }
        applySession(json)
        val rotated = json.optString("refresh_token", "")
        if (rotated.isNotEmpty()) {
            settings.podhopperRefreshToken.set(rotated, updateModifiedAt = false)
        }
        return cachedAccessToken ?: throw SupabaseException("PodHopper sync: missing access token")
    }

    private fun applySession(json: JSONObject) {
        cachedAccessToken = json.getString("access_token")
        val expiresInSec = json.optLong("expires_in", 3600)
        tokenExpiresAtMs = System.currentTimeMillis() + expiresInSec * 1000
        val user = json.optJSONObject("user")
        val userId = user?.optString("id").orEmpty()
        cachedUserId = if (userId.isEmpty()) null else userId
    }

    private fun clearSessionCache() {
        cachedAccessToken = null
        cachedUserId = null
        tokenExpiresAtMs = 0
    }

    private fun authCall(path: String, body: JSONObject): JSONObject {
        val request = Request.Builder()
            .url(PodHopperConfig.SUPABASE_URL + path)
            .addHeader("apikey", PodHopperConfig.SUPABASE_ANON_KEY)
            .post(body.toString().toRequestBody(jsonType))
            .build()
        httpClient.newCall(request).execute().use { response ->
            val responseBody = response.body.string()
            val code = response.code
            if (code == 400 || code == 401 || code == 403) {
                throw AuthRejectedException("PodHopper auth failed: HTTP $code $responseBody")
            }
            if (!response.isSuccessful) {
                throw SupabaseException("PodHopper auth failed: HTTP $code $responseBody")
            }
            return if (responseBody.isEmpty()) JSONObject() else JSONObject(responseBody)
        }
    }

    private fun restRequestBuilder(url: String): Request.Builder {
        val token = ensureSession()
        return Request.Builder()
            .url(url)
            .addHeader("apikey", PodHopperConfig.SUPABASE_ANON_KEY)
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
    }

    private fun executeExpectingSuccess(request: Request, retryOnAuthError: Boolean): String {
        httpClient.newCall(request).execute().use { response ->
            val responseBody = response.body.string()
            if (response.code == 401 && retryOnAuthError) {
                clearSessionCache()
                val refreshed = ensureSession()
                val retried = request.newBuilder()
                    .removeHeader("Authorization")
                    .addHeader("Authorization", "Bearer $refreshed")
                    .build()
                return executeExpectingSuccess(retried, retryOnAuthError = false)
            }
            if (!response.isSuccessful) {
                throw SupabaseException("PodHopper sync request failed: HTTP ${response.code} $responseBody")
            }
            return responseBody
        }
    }

    private companion object {
        const val TOKEN_SAFETY_MARGIN_MS = 60 * 1000L
    }
}

open class SupabaseException(message: String, cause: Throwable? = null) : Exception(message, cause)

class AuthRejectedException(message: String) : SupabaseException(message)
